### PR TITLE
[WaveTransform] Handle divergent PHIs with s1 def for GlobalIsel.

### PR DESCRIPTION
**Pull request by vg0204**
This change brings the GlobalISel divergence-lowering path under LWT a step closer to parity with the SDAG widening logic. Before widening divergent `s1` PHIs to `s32`, the pass now first folds any uniform PHI that merely merges a single real value through divergent control flow replacing it with that value directly instead of needlessly widening it.

Branch lowering (divergent `G_BRCOND` → `SI_BRCOND`) is split out of the phi widening routine into its own method for clarity; both still run back-to-back in the same pass. This is just a NFC-refactor.

Tests has been added for following coverage :
1. Uniform-PHI folding based on incoming edge values.
2. temporal divergence of a value used outside a loop.

### CODE REVIEW

**Comment by cdevadas** [comment]
There is an update in llvm/lib/Target/AMDGPU/AMDGPUGlobalISelDivergenceLowering.cpp with the latest merge. Rebase and resolve the conflicts.

File: llvm/test/CodeGen/AMDGPU/WaveTransform/gisel-divergent-i1-used-outside-loop.ll [hunk 1]
```diff
@@ -0,0 +1,242 @@
+; NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 6
+; RUN: llc -global-isel=1 -amdgpu-late-wave-transform=1 -mtriple=amdgcn-amd-amdpal -mcpu=gfx1100 -verify-machineinstrs -stop-after=amdgpu-global-isel-divergence-lowering < %s | FileCheck %s
```
Comment (cdevadas) [inline llvm/test/CodeGen/AMDGPU/WaveTransform/gisel-divergent-i1-used-outside-loop.ll]:
```suggestion
; RUN: llc -amdgpu-late-wave-transform=1 -global-isel=1 -mtriple=amdgpu11.00-amd-amdpal -verify-machineinstrs -stop-after=amdgpu-global-isel-divergence-lowering < %s | FileCheck %s
```

File: llvm/test/CodeGen/AMDGPU/WaveTransform/gisel-simplify-uniform-phi.ll [hunk 2]
```diff
@@ -0,0 +1,151 @@
+; NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 6
+; RUN: llc -global-isel=1 -amdgpu-late-wave-transform=1 -O0 -mtriple=amdgcn-amd-amdhsa -mcpu=gfx1010 -verify-machineinstrs -stop-after=amdgpu-global-isel-divergence-lowering < %s | FileCheck %s
```
Comment (cdevadas) [inline llvm/test/CodeGen/AMDGPU/WaveTransform/gisel-simplify-uniform-phi.ll]: Ditto.

File: llvm/lib/Target/AMDGPU/AMDGPUGlobalISelDivergenceLowering.cpp [hunk 3]
```diff
@@ -334,15 +337,85 @@ class DivergenceS1WideningHelper : public AMDGPU::PhiLoweringHelper {
...
+        if (MUI->isDivergentAtDef(In)) {
+          Bail = true;
+          break;
+        }
+
+        if (CommonReg && CommonReg != In) {
+          Bail = true;
+          break;
+        }
+        CommonReg = In;
```
Comment (cdevadas) [inline llvm/lib/Target/AMDGPU/AMDGPUGlobalISelDivergenceLowering.cpp]: This is required only when CommonReg is zero?

Comment (vg0204) [reply llvm/lib/Target/AMDGPU/AMDGPUGlobalISelDivergenceLowering.cpp]: No, its needed for any value as it represent the common register real value (if such exists or undef).

Comment (cdevadas) [reply llvm/lib/Target/AMDGPU/AMDGPUGlobalISelDivergenceLowering.cpp]: This code looks inefficient. It sets CommonReg = In when `CommonReg` already equals `In` (as the condition check indicates). So better change the code to something similar:
  if (!CommonReg)
    CommonReg = In;
  else if (CommonReg != In) {
    Bail = true;
    break;
  }

The same code style I see in AMDGPUFinalizeISelWaveTransform.cpp:
 if (CommonReg && CommonReg != Incoming)
    return Register();
  CommonReg = Incoming;
This should be corrected in a follow-up patch to match the same logic above.

File: llvm/lib/Target/AMDGPU/AMDGPUGlobalISelDivergenceLowering.cpp [hunk 4]
```diff
@@ -334,15 +337,86 @@ class DivergenceS1WideningHelper : public AMDGPU::PhiLoweringHelper {
...
+      bool Bail = false;
+
+      for (unsigned I = 1, E = MI.getNumOperands(); I < E; I += 2) {
+        Register In = MI.getOperand(I).getReg();
+        if (In == DstReg)
+          continue;
+
+        MachineInstr *Def = MRI->getVRegDef(In);
+        if (Def && (Def->isImplicitDef() ||
+                    Def->getOpcode() == TargetOpcode::G_IMPLICIT_DEF)) {
```
Comment (cdevadas) [inline llvm/lib/Target/AMDGPU/AMDGPUGlobalISelDivergenceLowering.cpp]: At this stage of GISel pipeline the implicit def can only be TargetOpcode::G_IMPLICIT_DEF. The other form of isImplicitDef() is unwanted? Can you check if any of the GISEL passes before actual instruction selection this form of implicit-def check is repeated.

File: llvm/lib/Target/AMDGPU/AMDGPUGlobalISelDivergenceLowering.cpp [hunk 5]
```diff
@@ -436,7 +516,10 @@ static bool runDivergenceLowering(MachineFunction &MF, MachineDominatorTree &DT,
   bool Changed = false;
   if (EnableLateWaveTransform) {
     MF.getInfo<SIMachineFunctionInfo>()->setWaveCFG(false);
-    Changed = DivergenceS1WideningHelper(MF, DT, PDT, &MUI).widenS1Phis();
+    DivergenceS1WideningHelper Helper(MF, DT, PDT, &MUI);
```
Comment (cdevadas) [inline llvm/lib/Target/AMDGPU/AMDGPUGlobalISelDivergenceLowering.cpp:518]: Technically, we could merge this handling alongside the DivergenceLoweringHelper. But for now, the separate handler seems reasonable. Let's see this later at our upstream effort.

**Review by cdevadas — APPROVED** [review]
LGTM except for the implicit-def check. Verify it once.

File: llvm/lib/Target/AMDGPU/AMDGPUGlobalISelDivergenceLowering.cpp (hunk 4 above)
Comment (vg0204) [reply llvm/lib/Target/AMDGPU/AMDGPUGlobalISelDivergenceLowering.cpp]: You are right to say that in GISel path, as at his point generic MIR will be used rather than post-Sel `IMPLICIT_DEF` (for which isImplicitDef() checks for), so it can be dropped off!
